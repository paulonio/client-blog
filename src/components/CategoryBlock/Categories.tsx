import React, { FC } from 'react';

import Link from 'next/link';
import { Category } from '@/types/types';
import { sen } from '@/styles/fonts';

import styles from './styled.module.scss';
import { parseString } from '@/utils/utils';

interface CategoriesProps {
  categories: Category[];
  type: 'home' | 'controls';
  currentCategory?: string;
}

const Categories: FC<CategoriesProps> = ({ categories, type, currentCategory }) => {
  if (!currentCategory) {
    return (
      <div className={`${styles.categories__wrapper} ${styles[type]}`}>
        {categories.map(({ title, description, Icon }) => (
          <Link key={`${Icon}`} href={`/${parseString(title)}`}>
            <div className={`${styles.category__wrapper} ${styles[type]}`}>
              <div className={`${styles.icon__wrapper} ${styles[type]}`}>
                <Icon />
              </div>
              <h3 className={sen.className}>{title}</h3>
              <p className="body-s">{description}</p>
            </div>
          </Link>
        ))}
      </div>
    );
  }

  return (
    <div className={`${styles.categories__wrapper} ${styles[type]}`}>
      {categories.map(({ title, Icon }) => {
        const category = parseString(currentCategory);
        const isActive = parseString(title) === category ? styles.active : '';
        return (
          <div
            key={`${Icon}`}
            className={`${styles.category__wrapper} ${isActive}  ${styles[type]}`}
          >
            <div className={`${styles.icon__wrapper} ${styles[type]}`}>
              <Icon />
            </div>
            <h3 className={sen.className}>{title}</h3>
          </div>
        );
      })}
    </div>
  );
};

export default Categories;
