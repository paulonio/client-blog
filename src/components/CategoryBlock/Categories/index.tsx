import React, { FC } from 'react';
import { useTranslation } from 'next-i18next';
import Link from 'next/link';

import { Category } from '@/types/types';
import { sen } from '@/styles/fonts';

import { parseString, stringToKey } from '@/utils/utils';
import styles from '../styled.module.scss';

interface CategoriesProps {
  categories: Category[];
  type: 'home' | 'controls';
  currentCategory?: string;
}

const Categories: FC<CategoriesProps> = ({ categories, type, currentCategory }) => {
  const { t } = useTranslation(['category', 'common']);

  return (
    <div className={`${styles.categories__wrapper} ${styles[type]}`}>
      {categories.map(({ title, description, Icon }) => {
        const category = currentCategory && parseString(currentCategory);
        const isActive = parseString(title) === category ? styles.active : '';
        return (
          <Link key={`${Icon}`} className={styles.link} href={`/${parseString(title)}`}>
            <div className={`${styles.category__wrapper} ${isActive}  ${styles[type]}`}>
              <div className={`${styles.icon__wrapper} ${styles[type]}`}>
                <Icon />
              </div>
              <h3 className={sen.className}>{t(stringToKey(title), { ns: 'common' })}</h3>
              {!category && <p className="body-s">{t(stringToKey(description))}</p>}
            </div>
          </Link>
        );
      })}
    </div>
  );
};

export default Categories;
