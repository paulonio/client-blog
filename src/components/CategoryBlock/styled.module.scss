@import '../../styles/sass/constants';

.categories {
  margin-top: 128px;
  display: flex;
  flex-direction: column;
  align-items: center;
  row-gap: 48px;

  &__wrapper {
    width: 100%;
    display: flex;
    align-items: center;
    justify-content: space-between;
    gap: 32px;

    @media screen and (max-width: 900px) {  
      flex-flow: row wrap;
      justify-content: center;
      gap: 24px;
    }

    &.controls {
      width: 100%;
      max-width: 296px;
      display: flex;
      flex-direction: column;
      align-items: stretch;
      row-gap: 24px;

      & a {
        width: 100%;
      }
    }
  }
}

.link {
  flex-shrink: 1;
}

.category__wrapper {
  padding: 32px;
  display: flex;
  flex-direction: column;
  align-items: start;
  row-gap: 4px;
  border: solid 1px $color-dark-s;
  transition: background-color 0.3s, border-color 0.3s;

  &.controls {
    width: 100%;
    padding: 23px;
    display: flex;
    flex-direction: row;
    align-items: center;
    gap: 16px;
    border: solid 1px $color-dark-s;
  }

  &.active {
    background-color: $color-primary;
    border: $color-primary;
  }

  &:hover {
    background-color: $color-primary;
    border-color: $color-primary;
    cursor: pointer;
  }

  @media screen and (max-width: 900px) {
    max-width: 315px;
  }
}

.icon__wrapper {
  width: 48px;
  height: 48px;
  margin-bottom: 12px;
  display: flex;
  align-items: center;
  justify-content: center;
  border-radius: 9px;
  background-color: $color-light-s;

  &.controls {
    width: 48px;
    height: 48px;
    margin-bottom: 0;
    display: flex;
    align-items: center;
    justify-content: center;
  }
}