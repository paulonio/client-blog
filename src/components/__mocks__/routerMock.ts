import { useRouter } from 'next/router';

jest.mock('next/router', () => ({
  useRouter: jest.fn(),
}));

(useRouter as jest.Mock).mockReturnValue({ query: {} });
