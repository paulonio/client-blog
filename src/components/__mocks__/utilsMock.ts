import { isLinkActive, parseString } from '@/utils/utils';

jest.mock('../../utils/utils', () => ({
  isLinkActive: jest.fn(),
  parseString: jest.fn(),
}));

(isLinkActive as jest.Mock).mockReturnValue(true);
(parseString as jest.Mock).mockReturnValue('');
