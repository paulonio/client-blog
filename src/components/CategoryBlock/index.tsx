'use client';

import React, { useRef } from 'react';
import { useTranslation } from 'next-i18next';

import Categories from './Categories';

import { CATEGORY_ITEMS } from '@/constants/constants';
import { useIntersectionObserver } from '@/hooks/useIntersectionObserver';

import { sen } from '@/styles/fonts';
import styles from './styled.module.scss';

const CategoryBlock = () => {
  const ref = useRef<HTMLElement | null>(null);
  const isIntersecting = useIntersectionObserver(ref);
  const { t } = useTranslation('category');

  return (
    <section ref={ref} className={styles.categories}>
      {isIntersecting && (
        <>
          <h2 className={sen.className}>{t('heading')}</h2>
          <Categories type="home" categories={CATEGORY_ITEMS} />
        </>
      )}
    </section>
  );
};

export default CategoryBlock;
