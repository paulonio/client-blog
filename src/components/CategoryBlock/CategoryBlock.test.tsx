import React from 'react';
import { render } from '@testing-library/react';
import CategoryBlock from '.';
import '../__mocks__/intersectionObserverMock';

describe('CategoryBlock', () => {
  it('should render category block', () => {
    const { container } = render(<CategoryBlock />);

    expect(container).toMatchSnapshot();
  });
});
