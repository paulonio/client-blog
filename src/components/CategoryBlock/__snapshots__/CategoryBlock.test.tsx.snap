// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CategoryBlock should render category block 1`] = `
<div>
  <section
    class="categories"
  />
</div>
`;
